jsonpb/jsonpb.go

```diff
@@ -825,13 +838,99 @@ func (u *Unmarshaler) unmarshalValue(target reflect.Value, inputValue json.RawMe
 		return nil
 	}
 
-	// 64-bit integers can be encoded as strings. In this case we drop
+	// Integers can be encoded as strings. In this case we drop
 	// the quotes and proceed as normal.
-	isNum := targetType.Kind() == reflect.Int64 || targetType.Kind() == reflect.Uint64
-	if isNum && strings.HasPrefix(string(inputValue), `"`) {
+	isInt := targetType.Kind() == reflect.Int64 || targetType.Kind() == reflect.Uint64 || target.Kind() == reflect.Int32 || target.Kind() == reflect.Uint32
+	if isInt && strings.HasPrefix(string(inputValue), `"`) {
 		inputValue = inputValue[1 : len(inputValue)-1]
 	}
 
+	// Floats have additional possible string values, Infinity, -Infinity, NaN -- handle those.
+	isFloat := targetType.Kind() == reflect.Float64 || targetType.Kind() == reflect.Float32
+	if isFloat && strings.HasPrefix(string(inputValue), `"`) {
+		reflectValueNaNFloat := map[reflect.Kind]reflect.Value{
+			reflect.Float32: reflect.ValueOf(float32(math.NaN())),
+			reflect.Float64: reflect.ValueOf(float64(math.NaN())),
+		}
+
+		reflectValuePosInfinityFloat := map[reflect.Kind]reflect.Value{
+			reflect.Float32: reflect.ValueOf(float32(math.Inf(1))),
+			reflect.Float64: reflect.ValueOf(float64(math.Inf(1))),
+		}
+
+		reflectValueNegInfinityFloat := map[reflect.Kind]reflect.Value{
+			reflect.Float32: reflect.ValueOf(float32(math.Inf(-1))),
+			reflect.Float64: reflect.ValueOf(float64(math.Inf(-1))),
+		}
+
+		switch string(inputValue) {
+		case `"NaN"`:
+			target.Set(reflectValueNaNFloat[targetType.Kind()])
+			return nil
+
+		case `"Infinity"`:
+			target.Set(reflectValuePosInfinityFloat[targetType.Kind()])
+			return nil
+
+		case `"-Infinity"`:
+			target.Set(reflectValueNegInfinityFloat[targetType.Kind()])
+			return nil
+
+		default:
+			inputValue = inputValue[1 : len(inputValue)-1]
+		}
+	}
+
+	if isInt || isFloat {
+		/*
+			var tmp float64
+			if err := json.Unmarshal(inputValue, &tmp); err != nil {
+				return err
+			}
+
+			switch targetType.Kind() {
+			case reflect.Uint32:
+				target.Set(reflect.ValueOf(uint32(tmp)))
+
+			case reflect.Uint64:
+				target.Set(reflect.ValueOf(uint64(tmp)))
+
+			case reflect.Int32:
+				target.Set(reflect.ValueOf(int32(tmp)))
+
+			case reflect.Int64:
+				target.Set(reflect.ValueOf(int64(tmp)))
+
+			case reflect.Float32:
+				target.Set(reflect.ValueOf(float32(tmp)))
+
+			case reflect.Float64:
+				target.Set(reflect.ValueOf(float64(tmp)))
+			}*/
+
+		var tmp float64
```

Comment:
Of note here: The use of float64 is causing loss of precision for int64 values. The proper solution would be an exponent parser that outputs int64s instead of floats, as the `strconv.ParseInt` doesn't handle exponents.